src/main/java/bio/terra/janitor/service/cleanup/FlightManager.java

```diff
@@ -0,0 +1,113 @@
+ * Stairway} to hand off work between the two.
+ *
+ * <p>This class is meant to be run by only the primary Janitor instance, not by the secondaries.
+ *
+ * <p>The handoff is done by this class and tracked in the cleanup_flight table and Stairway's
+ * database.
+ */
+public class FlightManager {
+  private Logger logger = LoggerFactory.getLogger(FlightManager.class);
+
+  /**
+   * A limit on how many flights to recover at startup. This is to prevent blowing out the service
+   * at start up, but if we have too many flights to recover, they will not all be recovered
+   * immediately.
+   */
+  private static final int RECOVERY_LIMIT = 1000;
+
+  private final Stairway stairway;
+  private final JanitorDao janitorDao;
+  private final FlightFactory cleanupFlightFactory;
+
+  public FlightManager(
+      Stairway stairway, JanitorDao janitorDao, FlightFactory cleanupFlightFactory) {
+    this.stairway = stairway;
+    this.janitorDao = janitorDao;
+    this.cleanupFlightFactory = cleanupFlightFactory;
+  }
+
+  /**
+   * Schedule a single resource for cleaning. Returns whether a flight id if a flight was attempted
+   * to be submitted to Stairway.
+   */
+  public Optional<String> submitFlight(Instant expiredBy) {
+    String flightId = stairway.createFlightId();
+    Optional<TrackedResource> resource = janitorDao.updateResourceForCleaning(expiredBy, flightId);
+    if (!resource.isPresent()) {
+      // No resource to schedule.
+      return Optional.empty();
+    }
+    submitToStairway(flightId, resource.get());
+    return Optional.of(flightId);
+  }
+
+  /**
+   * Recover tracked resources with flight ids in the Janitor's storage that are not known to
+   * Stairway. Resubmit the flights, returning how many flights were resubmitted.
+   *
+   * <p>This function assumes that it is not running concurrently with any other submissions to
+   * Stairway, e.g {@link #submitFlight(Instant)}.
+   */
+  public int recoverUnsubmittedFlights() {
+    List<JanitorDao.TrackedResourceAndFlight> resourceAndFlights =
+        janitorDao.retrieveResourcesWith(CleanupFlightState.INITIATING, RECOVERY_LIMIT);
+    if (resourceAndFlights.size() == RECOVERY_LIMIT) {
+      logger.error(
+          "Recovering as many flights as the limit {}. Some flights may still need recovering.",
+          RECOVERY_LIMIT);
+    }
+    int submissions = 0;
+    for (JanitorDao.TrackedResourceAndFlight resourceAndFlight : resourceAndFlights) {
+      String flightId = resourceAndFlight.cleanupFlight().flightId();
+      try {
+        // If there is some flight state, then the flight has been submitted successfully and does
+        // not need to be recovered. There's nothing to do but wait for the flight to update it's
+        // state from INITIATING.
+        // TODO(wchamber): add metrics.
+        stairway.getFlightState(flightId);
+      } catch (FlightNotFoundException e) {
```

Comment:
This is kind of expected behavior, I'm not sure what information is worth logging. If we did log something, it would be at info. I think the caller of this method may log how many flights were recovered total, but logging each flight isn't particularly interesting I think.